File: src/lib/seriesUtils.ts
import type { Series } from '../types';

function parseDateOnly(dateStr: string): Date {
  const [year, month, day] = dateStr.split('-').map(Number);
  return new Date(year, (month ?? 1) - 1, day ?? 1);
}

function isFutureAirDate(dateStr: string | null): boolean {
  if (!dateStr) return false;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return parseDateOnly(dateStr).getTime() > today.getTime();
}

function getReleasedSeasonCount(series: Series): number | null {
  if (series.seasons === null) return null;

  const hasFutureSeasonAnnounced =
    series.next_season_number !== null &&
    isFutureAirDate(series.next_air_date) &&
    series.next_season_number <= series.seasons;

  if (!hasFutureSeasonAnnounced) return series.seasons;

  return Math.max(0, (series.next_season_number ?? 1) - 1);
}

function hasWatchedAllPreviousEpisodesInNextSeason(series: Series): boolean {
  if (series.next_season_number === null || series.next_episode_number === null) return false;
  if (series.next_episode_number <= 1) return true;

  const watchedEpisodes = new Set(series.watched_episodes?.[String(series.next_season_number)] ?? []);
  for (let episode = 1; episode < series.next_episode_number; episode += 1) {
    if (!watchedEpisodes.has(episode)) return false;
  }

  return true;
}

export function deriveSeriesStatus(
  watchedSeasons: number[],
  totalSeasons: number | null,
  hasUnreleasedEpisodes = false,
  watchedEpisodes: Record<string, number[]> = {}
): 'watched' | 'watching' | 'want_to_watch' {
  const hasAnyEpisodeWatched =
    watchedSeasons.length > 0 || Object.values(watchedEpisodes).some(eps => eps.length > 0);
  if (!hasAnyEpisodeWatched) return 'want_to_watch';
  if (!hasUnreleasedEpisodes && totalSeasons && watchedSeasons.length >= totalSeasons) return 'watched';
  return 'watching';
}

export function getEffectiveSeriesStatus(series: Series): 'watched' | 'watching' | 'want_to_watch' {
  return deriveSeriesStatus(
    series.watched_seasons ?? [],
    series.seasons,
    false,
    series.watched_episodes ?? {}
  );
}

/**
 * Détermine si une série est en attente de nouveaux épisodes/saisons.
 * Une série est "en attente" si l'utilisateur a regardé toutes les saisons
 * disponibles et qu'une nouvelle saison/un nouvel épisode est attendu.
 */
export function isSeriesWaiting(s: Series): boolean {
  const effectiveStatus = getEffectiveSeriesStatus(s);
  if (effectiveStatus === 'want_to_watch') return false;
  if (!isFutureAirDate(s.next_air_date) || s.next_season_number === null) return false;
  if (effectiveStatus === 'watched') return true;

  if (s.next_episode_number !== null && s.next_episode_number > 1) {
    return hasWatchedAllPreviousEpisodesInNextSeason(s);
  }

  const nextSeasonKey = String(s.next_season_number);
  const hasStartedNextSeason =
    (s.watched_episodes?.[nextSeasonKey]?.length ?? 0) > 0 ||
    s.watched_seasons.includes(s.next_season_number);
  if (hasStartedNextSeason) return false;

  const releasedSeasonCount = getReleasedSeasonCount(s);
  if (releasedSeasonCount !== null && s.watched_seasons.length < releasedSeasonCount) return false;

  return s.next_season_number > s.watched_seasons.length;
}

/**
 * Formate le label du badge "En attente" avec la date si disponible.
 * - Pas de date : "En attente"
 * - Demain      : "demain"
 * - < 60 jours  : "dans Xj"
 * - >= 60 jours : "15 avr." / "avr. 2026"
 * - Date passée : "En attente" (la saison a probablement commencé)
 */
export function formatWaitingLabel(
  nextAirDate: string | null,
  fallback: string,
  tomorrowLabel: string,
  inDaysLabel: (days: number) => string,
  locale: string
): string {
  if (!nextAirDate) return fallback;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const date = new Date(nextAirDate);
  date.setHours(0, 0, 0, 0);
  const days = Math.round((date.getTime() - today.getTime()) / 86_400_000);
  if (days <= 0) return fallback;
  if (days === 1) return tomorrowLabel;
  if (days < 60) return inDaysLabel(days);
  // Même année → "15 avr.", sinon → "avr. 2026"
  if (date.getFullYear() === today.getFullYear()) {
    return date.toLocaleDateString(locale, { day: 'numeric', month: 'short' });
  }
  return date.toLocaleDateString(locale, { month: 'short', year: 'numeric' });
}

File: native/composeApp/src/commonMain/kotlin/fr/paulbr/nookmind/core/domain/SeriesUtils.kt
package fr.paulbr.nookmind.core.domain

import fr.paulbr.nookmind.core.model.Series
import fr.paulbr.nookmind.core.model.SeriesStatus
import kotlinx.datetime.LocalDate

// Port of src/lib/seriesUtils.ts — same rules, same edge cases (covered by SeriesUtilsTest).

private fun getReleasedSeasonCount(series: Series, today: LocalDate): Int? {
    val seasons = series.seasons ?: return null
    val nextSeason = series.nextSeasonNumber
    val hasFutureSeasonAnnounced =
        nextSeason != null && isFutureDate(series.nextAirDate, today) && nextSeason <= seasons
    if (!hasFutureSeasonAnnounced) return seasons
    return maxOf(0, (nextSeason ?: 1) - 1)
}

private fun hasWatchedAllPreviousEpisodesInNextSeason(series: Series): Boolean {
    val nextSeason = series.nextSeasonNumber ?: return false
    val nextEpisode = series.nextEpisodeNumber ?: return false
    if (nextEpisode <= 1) return true
    val watched = series.watchedEpisodes[nextSeason.toString()]?.toSet() ?: emptySet()
    for (episode in 1 until nextEpisode) {
        if (episode !in watched) return false
    }
    return true
}

fun deriveSeriesStatus(
    watchedSeasons: List<Int>,
    totalSeasons: Int?,
    hasUnreleasedEpisodes: Boolean = false,
    watchedEpisodes: Map<String, List<Int>> = emptyMap(),
): SeriesStatus {
    val hasAnyEpisodeWatched = watchedSeasons.isNotEmpty() || watchedEpisodes.values.any { it.isNotEmpty() }
    if (!hasAnyEpisodeWatched) return SeriesStatus.WANT_TO_WATCH
    if (!hasUnreleasedEpisodes && totalSeasons != null && totalSeasons > 0 && watchedSeasons.size >= totalSeasons) {
        return SeriesStatus.WATCHED
    }
    return SeriesStatus.WATCHING
}

fun getEffectiveSeriesStatus(series: Series, today: LocalDate = todayLocal()): SeriesStatus {
    val status = deriveSeriesStatus(series.watchedSeasons, series.seasons, false, series.watchedEpisodes)
    if (status == SeriesStatus.WATCHED && series.nextSeasonNumber != null && isFutureDate(series.nextAirDate, today)) {
        return SeriesStatus.WATCHING
    }
    return status
}

/**
 * A series is "waiting" when the user watched everything available and a new season /
 * episode is announced for a future date.
 */
fun isSeriesWaiting(s: Series, today: LocalDate = todayLocal()): Boolean {
    val effectiveStatus = getEffectiveSeriesStatus(s, today)
    if (effectiveStatus == SeriesStatus.WANT_TO_WATCH) return false
    val nextSeason = s.nextSeasonNumber
    if (!isFutureDate(s.nextAirDate, today) || nextSeason == null) return false
    if (effectiveStatus == SeriesStatus.WATCHED) return true

    val nextEpisode = s.nextEpisodeNumber
    if (nextEpisode != null && nextEpisode > 1) {
        if (nextSeason in s.watchedSeasons) return true
        return hasWatchedAllPreviousEpisodesInNextSeason(s)
    }

    val hasStartedNextSeason =
        (s.watchedEpisodes[nextSeason.toString()]?.size ?: 0) > 0 || nextSeason in s.watchedSeasons
    if (hasStartedNextSeason) return false

    val releasedSeasonCount = getReleasedSeasonCount(s, today)
    if (releasedSeasonCount != null && s.watchedSeasons.size < releasedSeasonCount) return false

    return nextSeason > s.watchedSeasons.size
}

/**
 * Label of the "waiting" badge:
 *  - no date → [fallback]
 *  - tomorrow → [tomorrowLabel]
 *  - < 60 days → [inDaysLabel]
 *  - ≥ 60 days → "15 avr." (same year) or "avr. 2026"
 *  - past date → [fallback]
 */
fun formatWaitingLabel(
    nextAirDate: String?,
    fallback: String,
    tomorrowLabel: String,
    inDaysLabel: (Int) -> String,
    formatDayMonth: (LocalDate) -> String,
    formatMonthYear: (LocalDate) -> String,
    today: LocalDate = todayLocal(),
): String {
    val date = parseDateOnly(nextAirDate) ?: return fallback
    val days = daysUntil(date, today)
    if (days <= 0) return fallback
    if (days == 1) return tomorrowLabel
    if (days < 60) return inDaysLabel(days)
    return if (date.year == today.year) formatDayMonth(date) else formatMonthYear(date)
}
